Add is_abs_ppt

sahilsingh2402 opened the pull request:
## Description
<!--Provide a brief description of the PR's purpose here. If your PR is supposed to fix an existing issue, use
a [keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) to link your PR to the issue. -->
This PR adds the is_abs_ppt function to check whether a quantum state is absolutely PPT (Positive Partial Transpose). It also includes unit tests to verify correctness.

## Changes
<!--Notable changes that this PR has either accomplished or will accomplish. Feel free to add more lines to the itemized list
below. -->
  -  Implemented is_abs_ppt function in toqito/state_props/is_abs_ppt.py.
  -  Added unit tests in toqito/state_props/tests/test_is_abs_ppt.py.
  -  Integrated function into toqito/state_props/__init__.py for module accessibility.
  -  Ensured compliance with ruff for code style and black for formatting.
  -  Referenced QETLAB's MATLAB implementation in the docstring.

## Checklist
Before marking your PR ready for review, make sure you checked the following locally. If this is your first PR, you might be notified of some workflow failures after a maintainer has approved the workflow jobs to be run on your PR. 

Additional information is available in the [documentation](https://toqito.readthedocs.io/en/latest/contributing.html#testing).

  -  [x] Use `ruff` for errors related to code style and formatting.
  -  [x] Verify all previous and newly added unit tests pass in `pytest`.
  -  [x] Check the documentation build does not lead to any failures. `Sphinx` build can be checked locally for any failures related to your PR
  -  [x] Use `linkcheck` to check for broken links in the documentation
  -  [x] Use `doctest` to verify the examples in the function docstrings work as expected.

sahilsingh2402 commented:
Hey @vprusso @purva-thakre I've submitted a pull request for adding the is_abs_ppt function along with test cases. When you have some time, could you please review it and let me know if any enhancements or improvements are needed?

vprusso commented on toqito/state_props/is_abs_ppt.py:
```diff
@@ -0,0 +1,128 @@
+from toqito.state_props.is_separable import is_separable
+
+
+def is_abs_ppt(rho: np.ndarray, dim: list[int] = None, max_constraints: int = 2612) -> int:
```
Let's slightly modify the return signature of this function such that it returns a value of type `bool | None` as opposed to one of type `int`. In the case where it is absolutely PPT, we would return `True`, not absolutely PPT return `False`, and inconclusive, return `None`.

vprusso commented on toqito/state_props/is_abs_ppt.py:
```diff
@@ -0,0 +1,128 @@
+    >>> from toqito.state_props import is_abs_ppt
+    >>> rho = np.array([[0.5, 0], [0, 0.5]])  # Maximally mixed state
+    >>> is_abs_ppt(rho, [2, 2])
+    1
```
Assuming that the type-signature comment above is changed, we would need to convert these examples as well.

vprusso commented on toqito/state_props/is_abs_ppt.py:
```diff
@@ -0,0 +1,128 @@
+
+    References
+    ==========
+    - MATLAB QETLAB Implementation: Nathaniel Johnston,
```
I think we may want to modify the `.bib` file and include a link to this function there. Then, we could just do:

```
    References
    ==========
    .. bibliography::
        :filter: docname in docnames
```

vprusso commented on toqito/state_props/is_abs_ppt.py:
```diff
@@ -0,0 +1,128 @@
+
+    eigvals = np.linalg.eigvalsh(rho)
+    p = int(min(dim))
+    print(f"[DEBUG] Computed p = {p}, type = {type(p)}")
```
Instead of printing out debug statements, we may instead want to use `logging.debug` here.

vprusso commented on toqito/state_props/is_abs_ppt.py:
```diff
@@ -0,0 +1,128 @@
+    p = int(min(dim))
+    print(f"[DEBUG] Computed p = {p}, type = {type(p)}")
+
+    if p >= 7:
```
In the case of `p >= 7`, we want to solve an SDP instead of just not computing anything. Refer to the QETLAB reference function:
https://github.com/nathanieljohnston/QETLAB/blob/0359eb36d73ebac1986879ec644dd684a1f62b6a/IsAbsPPT.m#L95

vprusso commented on toqito/state_props/is_abs_ppt.py:
```diff
@@ -0,0 +1,128 @@
+    """
+    rho = np.array(rho, dtype=np.complex128)
+
+    if rho.shape[0] != rho.shape[1]:
```
We can likely consolidate these two checks into one by using the `is_density_matrix` function in toqito

vprusso commented on toqito/state_props/is_abs_ppt.py:
```diff
@@ -0,0 +1,128 @@
+        return 0
+
+    try:
+        if is_separable(rho):
```
I'm wondering if checking `in_separable_ball` would be what we want here over `is_separable`?

vprusso reviewed (changes requested):
Thank you for your contribution, @sahilsingh2402 

Putting in a few comments, but feel free to push back or to ask for any clarifications on what I've written.

purva-thakre commented on toqito/state_props/is_abs_ppt.py:
```diff
@@ -0,0 +1,128 @@
+    - MATLAB QETLAB Implementation: Nathaniel Johnston,
+      "IsAbsPPT function" <http://www.qetlab.com/IsAbsPPT>
+
+    Parameters
+    ==========
+    rho : np.ndarray
+        The density matrix (or a vector of its eigenvalues) to be tested.
+    dim : list[int], optional
+        A list of two integers representing the dimensions of the bipartite subsystems.
+        If `None`, an equal split is assumed.
+    max_constraints : int, optional
+        The maximum number of SDP constraints to impose when using the symmetric extension hierarchy.
+
+    Returns
+    =======
+    int
+        - 1 if `rho` is absolutely PPT.
+        - 0 if `rho` is not absolutely PPT.
+        - -1 if the test is inconclusive (e.g., due to high dimension or computational constraints).
+
+    Raises
+    ======
+    ValueError
+        If `rho` is not square.
+    ValueError
+        If `rho` has zero trace.
+    ValueError
+        If `dim` does not match the shape of `rho`.
```
We have a different style for specifying params etc. 
https://toqito.readthedocs.io/en/latest/contributing.html#references-in-docstrings

purva-thakre commented:
@sahilsingh2402 Are you planning on finishing this?

sahilsingh2402 commented:
> @sahilsingh2402 Are you planning on finishing this?

Yeah, I will try to finish it by this weekend. Sorry I've been busy with some academics work of mine.

sahilsingh2402 commented:
Hi @purva-thakre  @vprusso, Whenever you get a chance, could you kindly take a look at the recent modifications I made? 

List of changes made:
- Updated return type from -1, 0, 1 to bool (True, False, None) for clarity and consistency.
- Adjusted all examples accordingly to reflect new return values.
- Removed .bib additions - used the existing .. bibliography:: directive as recommended.
- Removed all print/debug statements for cleanliness.
- Enabled proper SDP solving for p ≥ 7 instead of skipping .
- Replaced redundant rho.shape checks with is_density() from toqito.
- Used in_separable_ball instead of full is_separable check for efficiency.
- Cleaned up docstrings to follow Toqito’s style.

Please let me know if there's anything else you'd like me to change or improve. Thank you.

purva-thakre commented:
![image](https://github.com/user-attachments/assets/13b7875f-c503-41ae-a08e-157ef2e97822)

Resolve these conflicts.

sahilsingh2402 commented:
Hey, please check now. I have resolved the conflicts.

vprusso commented on toqito/state_props/is_abs_ppt.py:
```diff
@@ -0,0 +1,79 @@
+    lam = np.linalg.eigvalsh(rho)
+    p = min(dim)
+
+    if in_separable_ball(np.sort(lam)[::-1]):
+        return True
+    if not is_ppt(rho, sys=2, dim=dim):
+        return False
+
+    lam_sorted = np.sort(lam)[::-1]
+    if np.sum(lam_sorted[: p - 1]) <= 2 * lam_sorted[-1] + np.sum(lam_sorted[-(p - 1) : -1]):
+        return True
+
+    try:
+        L = symmetric_extension_hierarchy([rho], level=1, dim=dim, max_constraints=max_constraints)
+        if np.any(np.linalg.eigvals(L[-1]) < 0):
+            return False
+        return True if p <= 6 else None
+    except Exception:
+        return None
```
I'm not sure I understand how this section of the code relates to the absolutely PPT condition.

It seems like the check that you've imposed here are quite a bit different than those in:
https://github.com/nathanieljohnston/QETLAB/blob/0359eb36d73ebac1986879ec644dd684a1f62b6a/AbsPPTConstraints.m#L79

Can you explain how your logic here mimics what is done in the MATLAB reference? For instance, I don't see why the symmetric extension hierarchy is used in this case.

sahilsingh2402 replied:
> I'm not sure I understand how this section of the code relates to the absolutely PPT condition.
> 
> It seems like the check that you've imposed here are quite a bit different than those in: https://github.com/nathanieljohnston/QETLAB/blob/0359eb36d73ebac1986879ec644dd684a1f62b6a/AbsPPTConstraints.m#L79
> 
> Can you explain how your logic here mimics what is done in the MATLAB reference? For instance, I don't see why the symmetric extension hierarchy is used in this case.

Hey, sorry about the confusion earlier. I misunderstood the purpose of the check and wrongly assumed that using the symmetric_extension_hierarchy would be a valid fallback for higher-dimensional cases.

I’ve now updated the function to closely follow the logic from AbsPPTConstraints.m, including recursive matrix generation, ordering validation, and PSD checks on the constructed matrices, just as the original method requires.

vprusso commented on toqito/state_props/tests/test_is_abs_ppt.py:
```diff
@@ -0,0 +1,79 @@
+from toqito.states import bell, max_mixed
+
+
+@pytest.mark.parametrize(
```
Can you please verify that your approach also successfully runs on those in the reference implementation, please?
http://qetlab.com/AbsPPTConstraints

sahilsingh2402 commented:
Hi @purva-thakre @vprusso,  
Whenever you have a moment, I'd really appreciate it if you could take a look at the recent changes I made.

This time, I’ve implemented the functionality more directly by hardcoding the logic, earlier I attempted to use existing utilities from Toqito, but they didn’t align properly with the behavior required for absolute PPT checks.

I’ve now updated the implementation to closely follow the QETLAB version, including examples and test cases based on their original setup.

Please let me know if there’s anything else I should modify or improve. Thanks so much!

vprusso commented on toqito/state_props/tests/test_is_abs_ppt.py:
```diff
@@ -0,0 +1,69 @@
+            [2, 2],
+            True,
+        ),
+        # QETLAB examples
+        (np.eye(25) / 25, [5, 5], True),
+        (np.array([0.22, 0.18, 0.14, 0.12, 0.10, 0.08, 0.06, 0.06, 0.04]), [3, 3], False),
+        (np.array([0.3, 0.2, 0.15, 0.1, 0.08, 0.06, 0.04, 0.04, 0.03]), [3, 3], False),
+        (np.eye(49) / 49, [7, 7], None),
```
Can you tell me where you see these examples?

vprusso commented on toqito/state_props/is_abs_ppt.py:
```diff
@@ -0,0 +1,211 @@
+    if np.sum(lam[: p - 1]) <= 2 * lam[-1] + np.sum(lam[-p + 1 : -1]):
+        return True
+
+    if p >= 7:
```
I'm still not seeing the significance of "7" here. Why is this logic here?

vprusso commented on toqito/state_props/tests/test_is_abs_ppt.py:
```diff
@@ -0,0 +1,69 @@
+
+@pytest.mark.parametrize(
+    "rho, dim",
+    [
```
Can you test against the known examples of absolutely-PPT matrices? You can consult the "Examples" section of this paper for specific cases:
https://arxiv.org/pdf/quant-ph/0502170

vprusso commented on toqito/state_props/tests/test_is_abs_ppt.py:
```diff
@@ -0,0 +1,69 @@
+)
+def test_is_abs_ppt_known_values(rho, dim, expected_result):
+    """Test known examples and general functionality for is_abs_ppt."""
+    if callable(rho):
+        rho = rho()
```
When would `rho` be a callable?

vprusso commented on toqito/state_props/is_abs_ppt.py:
```diff
@@ -0,0 +1,211 @@
+    if not is_positive_semidefinite(constraints[-1]):
+        return False
+
+    if p <= 6:
+        return True
```
I'm very confused by this. Where does this come from?

sahilsingh2402 commented:
Closing this PR for now as per the suggestions.

purva-thakre commented:
Related to #208
